src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -1013,42 +1013,81 @@ string_exprt java_string_library_preprocesst::
 /*******************************************************************\
 
 Function: java_string_library_preprocesst::
-            code_assign_string_expr_to_java_string
+            code_assign_components_to_java_string
 
   Inputs:
-    lhs - an expression representing a java string
-    rhs - a string expression
+    lhs - expression representing the java string to assign to
+    rhs_array - pointer to the array to assign
+    rhs_length - length of the array to assign
     symbol_table - symbol table
 
   Output: return the following code:
-          > lhs->length=rhs.length
-          > lhs->data=&rhs.data
+          > lhs = { {Object}, length=rhs_length, data=rhs_array }
+
+ Purpose: Produce code for an assignemnrt of a string expr to a
+          Java string, component-wise.
 
 \*******************************************************************/
 
-codet java_string_library_preprocesst::code_assign_string_expr_to_java_string(
+codet java_string_library_preprocesst::code_assign_components_to_java_string(
   const exprt &lhs,
-  const string_exprt &rhs,
+  const exprt &rhs_array,
+  const exprt &rhs_length,
   symbol_tablet &symbol_table)
 {
   assert(implements_java_char_sequence(lhs.type()));
   dereference_exprt deref(lhs, lhs.type().subtype());
 
-  // Fields of the string object
-  exprt lhs_length=get_length(deref, symbol_table);
-  exprt lhs_data=get_data(deref, symbol_table);
-
-  // Assignments
   code_blockt code;
-  code.add(code_assignt(lhs_length, rhs.length()));
-  code.add(
-    code_assignt(lhs_data, address_of_exprt(rhs.content())));
+
+  // A String has a field Object with @clsid = String and @lock = false:
+  const symbolt &jlo_symbol=symbol_table.lookup("java::java.lang.Object");
+  const struct_typet &jlo_struct=to_struct_type(jlo_symbol.type);
+  struct_exprt jlo_init(jlo_struct);
+  jlo_init.copy_to_operands(constant_exprt(
+    "java::java.lang.String", jlo_struct.components()[0].type()));
```

Comment:
Recommend using https://github.com/diffblue/cbmc/blob/master/src/java_bytecode/java_object_factory.cpp#L264 to do this, to avoid duplicating implementations of java.lang.Object init and automatically keep up to date if its format changes.